### PR TITLE
Async capability combinators

### PR DESCRIPTION
**Pull request by charypar**
This PR is work in progress on #91.

It turns out, surprisingly, that async capability combinators were always possible in "user land", and could be implemented as an additional capability, but the implementation is not exactly obvious, and this type of use needs to be supported by the capabilities being orchestrated across, especially `crux_http`.

This PR aims to make the use of `async` across multiple capabilities more obvious. There are a few challenges to overcome

- [x] The Orchestrate capability should probably come with crux_core
- [ ] ~Automatically derive `Clone` for capabilities and the `Capabilities` struct, to avoid various lifetime errors when moving the capabilities or the context into the async blocks~  - decided against, it would conflict with custom `Clone` implementations
- [x] Provide a `Never` operation type
- [x] Avoid producing an `Effect` variant for capabilities who's ~operations are `Never` ()~ fields are `#[effect(skip)]`
- [x] The use needs thorough documentation and an example
- [ ] Existing capabilities need to expose an async version of their APIs (#180, #181, #182)

All feedback on the ergonomics of this is welcome at this stage, but I'd like to aim to ship this soon, ideally this week.

### CODE REVIEW

**Comment by wasnotrice** [comment]
@charypar thanks for this, it looks great! I'm so excited to try this in my app. I think it will help a lot with maintainability. I have some places where the flow of orchestrated events is really hard to follow.

In the example, there are two concurrent events, but it could easily be two sequential events, or any other orchestration, right?

**Comment by charypar** [comment]
@wasnotrice Yea, absolutely, it's just an open ended async context with access to the capability runtime, I'm pretty sure it can orchestrate an arbitrary graph of dependent effects (and emit events alon the way meantime). 

One thing to be aware off is that it is likely to be quite hard to test "sub-steps" in isolation on long effect chains. The entire outer future is one atomic flow which can only be entered into at the top.

File: crux_core/src/capabilities/compose.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    ///
+    /// The argument is a closure which receives a [`ComposeContext`] which can be used to send
+    /// events to the app.
+    ///
+    /// For example:
+    /// ```
+    /// fn update(&self, event: Self::Event, model: &mut Self::Model, caps: &Self::Capabilities) {
+    ///     match event {
+    ///        Event::Trigger => caps.compose.spawn(|context| {
+    ///             let caps = Clone::clone(caps);
```
Comment (obmarg) [inline crux_core/src/capabilities/compose.rs:55]: Not a big deal, but maybe: 

```suggestion
    ///             let caps = caps.clone();
```

Don't think I've ever seen someone do `Clone::clone`, looks a bit odd

Comment (obmarg) [reply crux_core/src/capabilities/compose.rs:55]: > @StuartHarris pointed out that Clone::clone is somewhat more explicit of the intent because it doesn't rely on automatic dereferencing.

True, though I'd usually express that as `WhateverTheActualTypeIsCalled::clone`.  Although maybe that's just me.

But whatever, does not matter.

Comment (charypar) [reply crux_core/src/capabilities/compose.rs:55]: Ah yes. I might still change it to `.clone()` because it's the least likely to trip the reader up and make them wonder what is going on.

File: crux_macros/src/effect.rs [hunk 2]
```diff
@@ -15,9 +15,18 @@ struct EffectStructReceiver {
 }
 
 #[derive(FromField, Debug)]
+#[darling(attributes(effect))]
 pub struct EffectFieldReceiver {
     ident: Option<Ident>,
     ty: Type,
+    skip: Option<bool>,
```
Comment (obmarg) [inline crux_macros/src/effect.rs:22]:
```suggestion
    #[darling(default)]
    skip: bool,
```

**Review by StuartHarris — APPROVED** [review]
yay!
